theano/tensor/opt.py

```diff
@@ -242,27 +250,36 @@ def inplace_elemwise_optimizer(fgraph):
             # If big graph and the outputs are scalar, do not make it
             # inplace.
             if (check_each_change != 1 and
-                all([getattr(o.type, 'ndim', -1) == 0
-                     for o in node.outputs])):
+                # If multiple outputs, they must all have the same size,
+                # so only check the first.
+                    getattr(node.outputs[0].type, 'ndim', -1) == 0):
                 continue
 
-            baseline = op.inplace_pattern
-            protected_inputs = [
-                f.protected for f in node.fgraph._features if
-                isinstance(f, theano.compile.function_module.Supervisor)]
-            protected_inputs = sum(protected_inputs, [])  # flatten the list
-            protected_inputs.extend(fgraph.outputs)
-            candidate_outputs = [i for i in xrange(len(node.outputs))
-                                 if i not in baseline]
-            # node inputs that are Constant, already destroyed,
-            # fgraph protected inputs and fgraph outputs can't be used as inplace
-            # target.
-            # Remove here as faster.
-            candidate_inputs = [i for i in xrange(len(node.inputs))
-                                if i not in baseline.values() and
-                                not isinstance(node.inputs[i], Constant) and
-                                not fgraph.destroyers(node.inputs[i]) and
-                                node.inputs[i] not in protected_inputs]
+            if op.inplace_pattern:
+                baseline = op.inplace_pattern
+                candidate_outputs = [i for i in xrange(len(node.outputs))
+                                     if i not in baseline]
+                # node inputs that are Constant, already destroyed,
+                # or fgraph protected inputs and fgraph outputs can't be used as
+                # inplace target.
+                # Remove here as faster.
+                candidate_inputs = [i for i in xrange(len(node.inputs))
+                                    if i not in baseline.values() and
+                                    not isinstance(node.inputs[i], Constant) and
+                                    # Is next line costly?
+                                    not fgraph.destroyers(node.inputs[i]) and
+                                    node.inputs[i] not in protected_inputs]
+            else:
+                baseline = []
```

Comment:
it is used here:

https://github.com/nouiz/Theano/blob/1328ebb3a208c400eb5ab2424cfdd8caa59d8805/theano/tensor/opt.py#L347

On Wed, May 11, 2016 at 1:21 PM, abergeron <notifications@github.com> wrote:

> In theano/tensor/opt.py
> <https://github.com/Theano/Theano/pull/4496#discussion_r62889211>:
>
> > +            if op.inplace_pattern:
> > +                baseline = op.inplace_pattern
> > +                candidate_outputs = [i for i in xrange(len(node.outputs))
> > +                                     if i not in baseline]
> > +                # node inputs that are Constant, already destroyed,
> > +                # or fgraph protected inputs and fgraph outputs can't be used as
> > +                # inplace target.
> > +                # Remove here as faster.
> > +                candidate_inputs = [i for i in xrange(len(node.inputs))
> > +                                    if i not in baseline.values() and
> > +                                    not isinstance(node.inputs[i], Constant) and
> > +                                    # Is next line costly?
> > +                                    not fgraph.destroyers(node.inputs[i]) and
> > +                                    node.inputs[i] not in protected_inputs]
> > +            else:
> > +                baseline = []
>
> This doesn't look like it's used anywhere.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4496/files/1328ebb3a208c400eb5ab2424cfdd8caa59d8805#r62889211>
>